### PR TITLE
990 suspend job

### PR DESCRIPTION
**Pull request by ggarri**
## PR description
- new supported status for jobs: SUSPENDED
- tx-listener suspends jobs on errors, except Dependency and EthClient error
- Added new transaction endpoint to resume suspended transactions
- Users can define a limit on tx listening based on the number of blocks, by default 30 blocks
- Removed db constraint job status values

## Fixed Issue(s)
#990 

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://github.com/consensys/orchestrate/blob/master/CHANGELOG.md).

### CODE REVIEW

File: src/tx-listener/tx-listener/sessions/jobs/session.go [hunk 1]
```diff
@@ -104,29 +119,35 @@ func (ses *SessionManager) runSession(ctx context.Context, job *entities.Job, ch
 		retryTicker.Stop()
 	}
 
-	// @TODO Add finality to suspend the job
+	jobListeningUntil := job.InternalData.SentAt.Add(
+		time.Duration(chain.ListenerTransactionWaitingBlocks) * chain.ListenerBlockTimeDuration)
```
Comment (darioAnongba) [inline src/tx-listener/tx-listener/sessions/jobs/session.go:123]: how can this work if the `ListenerTransactionWaitingBlocks` is a uint64? :thinking:

File: src/tx-sender/tx-sender/use-cases/sender/send-eth.go [hunk 2]
```diff
@@ -118,6 +118,7 @@ func (uc *sendETHTxUseCase) sendTx(ctx context.Context, job *entities.Job) (*eth
 	proxyURL := client.GetProxyURL(uc.chainRegistryURL, job.ChainUUID)
 	txHash, err := uc.ec.SendRawTransaction(ctx, proxyURL, job.Transaction.Raw)
 	if err != nil {
+		// @TODO AlreadyKnownTx in case of resending
```
Comment (darioAnongba) [inline src/tx-sender/tx-sender/use-cases/sender/send-eth.go:121]: Did you check this?

**Review by darioAnongba — APPROVED** [review]

File: src/tx-listener/tx-listener/use-cases/jobs/suspend.go [hunk 3]
```diff
@@ -12,30 +12,29 @@ import (
...
+	return &suspendJobUseCase{
 		pendingJobState: pendingJobState,
 		messenger:       messengerCli,
 		logger:          logger.SetComponent(failedSessionJobComponent),
 	}
 }
 
 // Execute starts a job session
-func (uc *failedJobUseCase) Execute(ctx context.Context, job *entities.Job, errStr string) error {
+func (uc *suspendJobUseCase) Execute(ctx context.Context, job *entities.Job, errStr string) error {
```
[the remark this answers is not in the document]
Comment (ggarri) [reply src/tx-listener/tx-listener/use-cases/jobs/suspend.go:32]: It is not only the messenger, but also to remove the local state of the pending job. Same as it used to do for the failed transaction before.

File: src/api/service/controllers/transactions.go [hunk 4]
```diff
@@ -311,6 +313,31 @@ func (c *TransactionsController) speedUp(rw http.ResponseWriter, req *http.Reque
...
+// @Param        uuid   path      string               true   "UUID of the transaction request"
+// @Failure      404    {object}  infra.ErrorResponse  "Transaction request not found"
+// @Failure      500    {object}  infra.ErrorResponse  "Internal server error"
+// @Router       /transactions/{uuid}/resume [put]
+func (c *TransactionsController) resume(rw http.ResponseWriter, req *http.Request) {
+	rw.Header().Set("Content-Type", "application/json")
+	ctx := req.Context()
+	var err error
+
+	uuid := mux.Vars(req)["uuid"]
```
[the remark this answers is not in the document]
Comment (ggarri) [reply src/api/service/controllers/transactions.go:331]: done

File: src/api/business/use-cases/jobs/create_job.go [hunk 5]
```diff
@@ -37,12 +36,15 @@ func (uc *createJobUseCase) Execute(ctx context.Context, job *entities.Job, user
...
 	logger.Debug("creating new job")
 
-	chainID, err := uc.getChainID(ctx, job.ChainUUID, userInfo)
+	chain, err := uc.getChainUC.Execute(ctx, job.ChainUUID, userInfo)
 	if err != nil {
-		return nil, errors.FromError(err).ExtendComponent(createJobComponent)
+		if errors.IsNotFoundError(err) {
+			return nil, errors.InvalidParameterError("failed to get chain")
+		}
+		return nil, errors.FromError(err)
```
[the remark this answers is not in the document]
Comment (ggarri) [reply src/api/business/use-cases/jobs/create_job.go:44]: done

File: src/api/business/use-cases/transactions/resume_tx.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+func NewResumeTxUseCase(getTxUC usecases.GetTxUseCase, resendJobTxUC usecases.ResendJobTxUseCase) usecases.ResumeTxUseCase {
+	return &resumeTxUseCase{
+		getTxUC:       getTxUC,
+		resendJobTxUC: resendJobTxUC,
+		logger:        log.NewLogger().SetComponent("use-cases.resume-tx"),
+	}
+}
+
+func (uc *resumeTxUseCase) Execute(ctx context.Context, scheduleUUID string, userInfo *multitenancy.UserInfo) (*entities.TxRequest, error) {
```
[the remark this answers is not in the document]
Comment (ggarri) [reply src/api/business/use-cases/transactions/resume_tx.go:27]: good catch. Added

File: src/api/service/types/chain.go [hunk 7]
```diff
@@ -15,8 +15,9 @@ type RegisterChainRequest struct {
 }
 
 type RegisterListenerRequest struct {
-	Depth             uint64 `json:"depth,omitempty" example:"0"`                                              // Block depth after which the Transaction Listener considers a block final and processes it (default 0).
-	BlockTimeDuration string `json:"blockTimeDuration,omitempty" validate:"omitempty,isDuration" example:"1s"` // Time to wait before trying to fetch a new mined block (for example `1s` or `1m`, default is `5s`).
+	Depth                    uint64 `json:"depth,omitempty" example:"0"`                                                // Block depth after which the Transaction Listener considers a block final and processes it (default 0).
+	BlockTimeDuration        string `json:"blockTimeDuration,omitempty" validate:"omitempty,isDuration" example:"10s"`  // Time to wait before trying to fetch a new mined block (for example `1s` or `1m`, default is `10s`).
+	TransactionWaitingBlocks uint64 `json:"transactionWaitingBlocks,omitempty" validate:"omitempty,min=1" example:"30"` // Max amount of block waited for giving up on transaction receipt (for example `30`, default is `30`).
```
[the remark this answers is not in the document]
Comment (ggarri) [reply src/api/service/types/chain.go]: using transaction time duration as discussed

File: src/tx-listener/tx-listener/sessions/jobs/session.go (hunk 1 above)
Comment (ggarri) [reply src/tx-listener/tx-listener/sessions/jobs/session.go]: using duration as discussed

File: src/tx-sender/tx-sender/use-cases/sender/send-eth.go (hunk 2 above)
Comment (ggarri) [reply src/tx-sender/tx-sender/use-cases/sender/send-eth.go:121]: I did, and it ignores the resend if tx is still not mined, but once it mined it fails with a "nonce too low" which is trait as "expected error" and ignored in the tx-sender

**Pull request merged**